gatsby/content/docs/2020-04-14-stateresv2.mdx

```diff
@@ -0,0 +1,329 @@
+
+```
+var resolved_power_level_event: partial_state["m.room.power_level"][""] as event
+var power_level_mainline:       list(event)
+
+func mainline_iterate(e as event):
+    power_level_mainline += e
+    for each e.auth_events as auth_event:
+        if auth_event.event_type == "m.room.power_level":
+            mainline_iterate(auth_event)
+```
+
+We refer to the time period between any two given power level events as an _epoch_.
+
+In the below illustration, the "resolved" event is the event that appears in the partial state from the previous section.
+
+![drawing showing partial state example](/docs/guides/img/2020-04-14-stateresv2-2.svg)
+
+### Mainline ordering of normal state events
+
+Since the mainline will be ordered based on when the power level changes happened, and other state events use power level events as auth events, it is possible to sort the order of normal state events based on which mainline power level event they most closely refer to, or by definition, which epoch the event happened in.
+
+Note that events may appear that refer to power level events which don't appear in the mainline, in which case it is necessary to step through the power level events iteratively until you encounter one that is on the mainline. This is referred to as the _closest mainline event_:
+
+```
+func is_in_mainline(e as event) returns boolean:
+    if e in power_level_mainline:
+        return true
+    else:
+        return false
+
+func get_closest_mainline_event(e as event) returns event:
+    var closest_mainline_event: event
+
+    func closest_iterate(e as event):
+    if is_in_mainline(e):
+            closest_mainline_event = e
+        else:
+            for each e.auth_events as auth_event:
+                if auth_event.event_type == "m.room.power_level":
+                    closest_iterate(auth_event)
+
+    closest_iterate(e)
+    return closest_mainline_event
+```
+
+In the following illustration, _Event B_ happens in a more recent epoch than _Event A_ as the closest mainline event referred to by _Event A_ is older than the closest mainline event referred to by _Event B_:
+
+![drawing showing an example of mainline ordering](/docs/guides/img/2020-04-14-stateresv2-3.svg)
+
+To sort the events, we use a method very similar to that used in the reverse topological ordering of control events, only instead of initially sorting by power level, we use the position of the nearest mainline power level event on the mainline. When sorting, you can assert that **x &lt; y** by running the following tests in order:
+
+1. The closest mainline event to **x** is later in the mainline then the closest mainline event to **y** (for example, in the above diagram, **A** is less than **B **as **B **refers to a more recent mainline power level event)
```

Comment:
> The closest mainline event to **x** is later in the mainline then the closest mainline event to **y**

So `later` is defined as being older? It's just I'd imagine `later` being further along in the event list, but we order the list such that older events are at the beginning of the list.

Reply:
`later` is newer. Maybe that's ambiguous?